<a href="https://colab.research.google.com/github/Mikaelow/Kaggle/blob/main/car_predicting_price.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [11]:
# importing librares
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)


In [12]:
# uploading dataset
db_source = 'cars_formated.csv'
# puting data base to the
cars_formated = pd.read_csv(db_source, index_col = 'Unnamed: 0').sample(50000,random_state=1)

In [13]:
cars_formated

year       make          model           body transmission  \
234624  2014      dodge        journey            suv    automatic   
99107   2011     toyota     highlander            suv    automatic   
252966  2007     toyota          camry          sedan    automatic   
383114  2012     nissan     pathfinder            suv    automatic   
159656  2013     nissan         altima          sedan    automatic   
...      ...        ...            ...            ...          ...   
352441  2012     toyota     highlander            suv    automatic   
426376  2013    hyundai  genesis coupe  genesis coupe    automatic   
510213  2008      buick        enclave            suv    automatic   
385790  2014     nissan          versa          sedan    automatic   
153707  2010  chevrolet          tahoe            suv    automatic   

                      vin  condition  odometer   color  sellingprice  
234624  3c4pdcbb3et180538        4.0   20366.0     red       15700.0  
99107   5tddk3eh1bs050093       36.0   97494.0   black       21500.0  
252966  4t1be46k47u078676       37.0   93568.0     red        9200.0  
383114  5n1ar1nn0cc627705       29.0   54826.0   black       15900.0  
159656  1n4al3ap8dn515075        4.0   60131.0  silver       12400.0  
...                   ...        ...       ...     ...           ...  
352441  5tddk3eh9cs152792       41.0   31666.0   black       30800.0  
426376  kmhhu6kj5du104607        4.0   25726.0   white       21000.0  
510213  5gaer13708j292332       42.0  112777.0   brown       11700.0  
385790  3n1cn7ap7el817613       43.0   37363.0   white       10000.0  
153707  1gnukce0xar286837       45.0   99681.0   white       24500.0  

[50000 rows x 10 columns]

In [14]:
# choosing most value columns from the data base
columns_choosen = cars_formated.columns
columns_choosen = columns_choosen.drop('vin')

In [15]:
cars_formated.dtypes

year              int64
make             object
model            object
body             object
transmission     object
vin              object
condition       float64
odometer        float64
color            object
sellingprice    float64
dtype: object

In [16]:
# encoding from object to bool
one_hot_encoded_db = pd.get_dummies(cars_formated[columns_choosen])

one_hot_encoded_db

year  condition  odometer  sellingprice  make_acura  \
234624  2014        4.0   20366.0       15700.0       False   
99107   2011       36.0   97494.0       21500.0       False   
252966  2007       37.0   93568.0        9200.0       False   
383114  2012       29.0   54826.0       15900.0       False   
159656  2013        4.0   60131.0       12400.0       False   
...      ...        ...       ...           ...         ...   
352441  2012       41.0   31666.0       30800.0       False   
426376  2013        4.0   25726.0       21000.0       False   
510213  2008       42.0  112777.0       11700.0       False   
385790  2014       43.0   37363.0       10000.0       False   
153707  2010       45.0   99681.0       24500.0       False   

        make_aston martin  make_audi  make_bentley  make_bmw  make_buick  ...  \
234624              False      False         False     False       False  ...   
99107               False      False         False     False       False  ...   
252966              False      False         False     False       False  ...   
383114              False      False         False     False       False  ...   
159656              False      False         False     False       False  ...   
...                   ...        ...           ...       ...         ...  ...   
352441              False      False         False     False       False  ...   
426376              False      False         False     False       False  ...   
510213              False      False         False     False        True  ...   
385790              False      False         False     False       False  ...   
153707              False      False         False     False       False  ...   

        color_lime  color_off-white  color_orange  color_pink  color_purple  \
234624       False            False         False       False         False   
99107        False            False         False       False         False   
252966       False            False         False       False         False   
383114       False            False         False       False         False   
159656       False            False         False       False         False   
...            ...              ...           ...         ...           ...   
352441       False            False         False       False         False   
426376       False            False         False       False         False   
510213       False            False         False       False         False   
385790       False            False         False       False         False   
153707       False            False         False       False         False   

        color_red  color_silver  color_turquoise  color_white  color_yellow  
234624       True         False            False        False         False  
99107       False         False            False        False         False  
252966       True         False            False        False         False  
383114      False         False            False        False         False  
159656      False          True            False        False         False  
...           ...           ...              ...          ...           ...  
352441      False         False            False        False         False  
426376      False         False            False         True         False  
510213      False         False            False        False         False  
385790      False         False            False         True         False  
153707      False         False            False         True         False  

[50000 rows x 768 columns]

# Spliting data to features and labels

In [17]:
features = one_hot_encoded_db.columns
# spliting to feature data base
# X = cars_formated[features]
features = features.drop('sellingprice')
features

Index(['year', 'condition', 'odometer', 'make_acura', 'make_aston martin',
       'make_audi', 'make_bentley', 'make_bmw', 'make_buick', 'make_cadillac',
       ...
       'color_lime', 'color_off-white', 'color_orange', 'color_pink',
       'color_purple', 'color_red', 'color_silver', 'color_turquoise',
       'color_white', 'color_yellow'],
      dtype='object', length=767)

In [18]:
# spliting to label
X = one_hot_encoded_db[features]
y = one_hot_encoded_db['sellingprice']
X

year  condition  odometer  make_acura  make_aston martin  make_audi  \
234624  2014        4.0   20366.0       False              False      False   
99107   2011       36.0   97494.0       False              False      False   
252966  2007       37.0   93568.0       False              False      False   
383114  2012       29.0   54826.0       False              False      False   
159656  2013        4.0   60131.0       False              False      False   
...      ...        ...       ...         ...                ...        ...   
352441  2012       41.0   31666.0       False              False      False   
426376  2013        4.0   25726.0       False              False      False   
510213  2008       42.0  112777.0       False              False      False   
385790  2014       43.0   37363.0       False              False      False   
153707  2010       45.0   99681.0       False              False      False   

        make_bentley  make_bmw  make_buick  make_cadillac  ...  color_lime  \
234624         False     False       False          False  ...       False   
99107          False     False       False          False  ...       False   
252966         False     False       False          False  ...       False   
383114         False     False       False          False  ...       False   
159656         False     False       False          False  ...       False   
...              ...       ...         ...            ...  ...         ...   
352441         False     False       False          False  ...       False   
426376         False     False       False          False  ...       False   
510213         False     False        True          False  ...       False   
385790         False     False       False          False  ...       False   
153707         False     False       False          False  ...       False   

        color_off-white  color_orange  color_pink  color_purple  color_red  \
234624            False         False       False         False       True   
99107             False         False       False         False      False   
252966            False         False       False         False       True   
383114            False         False       False         False      False   
159656            False         False       False         False      False   
...                 ...           ...         ...           ...        ...   
352441            False         False       False         False      False   
426376            False         False       False         False      False   
510213            False         False       False         False      False   
385790            False         False       False         False      False   
153707            False         False       False         False      False   

        color_silver  color_turquoise  color_white  color_yellow  
234624         False            False        False         False  
99107          False            False        False         False  
252966         False            False        False         False  
383114         False            False        False         False  
159656          True            False        False         False  
...              ...              ...          ...           ...  
352441         False            False        False         False  
426376         False            False         True         False  
510213         False            False        False         False  
385790         False            False         True         False  
153707         False            False         True         False  

[50000 rows x 767 columns]

# Building first model

In [19]:
# importing models
from sklearn.tree import DecisionTreeRegressor
# import mae
from sklearn.metrics import mean_absolute_error
# import db split to validate data
from sklearn.model_selection import train_test_split

In [20]:
# define model. Specify a number for random_state to ensure same results each run
car_model = DecisionTreeRegressor(random_state=1)

# fit model
car_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

In [21]:
# predicting price
pred_car_price = car_model.predict(X)
# checking mae
mae_pred = mean_absolute_error(y,pred_car_price)
mae_pred

0.0

In [22]:
# % of the mean
mae_pred_proc = (mae_pred/y.mean())*100
mae_pred_proc

0.0

**Observation: MAE is very low, only `0`, it represents just `0%` of the average car price.**

# Validating model

In [23]:
# building new model

# spliting data to train and validation
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)
# define new car model
car_model_with_validation = DecisionTreeRegressor(random_state = 1)
# fiting model
car_model_with_validation.fit(train_X, train_y)
# predicting price
pred_car_price_with_validation = car_model_with_validation.predict(val_X)

In [24]:
# checking mae
mae_pred_with_validation = mean_absolute_error(val_y, pred_car_price_with_validation)
mae_pred_with_validation

2491.61416

In [25]:
# % of the mean
mae_pred_with_validation_proc = (mae_pred_with_validation/y.mean())*100
mae_pred_with_validation_proc

18.200694691059923

**Observation: MAE is much bigger than without validation `2491.61`, it represents just `18.20%` of the average car price.**

# Underfitting and Overfitting

In [26]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state = 1)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [27]:
# compare MAE with differing values of max_leaf_nodes
for max_leaf_nodes in [5, 50, 500, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error:  4904
Max leaf nodes: 50  		 Mean Absolute Error:  3797
Max leaf nodes: 500  		 Mean Absolute Error:  2734
Max leaf nodes: 5000  		 Mean Absolute Error:  2287


In [28]:
my_mae_proc = (my_mae  /y.mean())*100
my_mae_proc

16.71292570389179

**Observation: From the `max_leaf_nodes = [5, 50, 500, 5000]`, the best MAE occurs when `max_leaf_nodes` is equal to `5000`, with a value of `2287`, representing just `16.72%`.**

# Creating random forest

In [29]:
# importing random forest
from sklearn.ensemble import RandomForestRegressor

# define random model
forest_model = RandomForestRegressor(random_state=1)
# fiting model
forest_model.fit(train_X, train_y)
# predicting
car_price_forest_pred = forest_model.predict(val_X)
# compare car_price_forest_pred with val_y in mae
mae_random_forest = mean_absolute_error(val_y, car_price_forest_pred)
mae_random_forest

In [37]:
mae_random_forest_proc = (mae_random_forest/y.mean())*100
mae_random_forest_proc

13.725820841040933

**Observation: RandomForestRegressor have the best from previous method MAE equal `1879.02`, representing just `13.73%`.**

# Create Linear Regression

In [31]:
# importing Linear Regression
from sklearn.linear_model import LinearRegression

# creating Linear Regression model
linear_regression_model = LinearRegression()

# fiting model
linear_regression_model.fit(train_X, train_y)

# predicting price
pred_price_linear_regression = linear_regression_model.predict(val_X)

# compare car_price_forest_pred with val_y in mae
mae_linear_regression = mean_absolute_error(val_y, pred_price_linear_regression)
mae_linear_regression

2709.3145231385793

In [32]:
mae_linear_regression_proc = (mae_linear_regression/y.mean())*100
mae_linear_regression_proc

19.790948072674254

**Observation: RandomForestRegressor have the best from previous method MAE equal `2709.31`, representing just `19.79%`.**

# Create SVM

In [33]:
# importing SVM
from sklearn import svm

# creating model
svm_model = svm.SVC()

# fiting model
svm_model.fit(train_X, train_y)

# predicting price
pred_price_svm = svm_model.predict(val_X)

#compare pred_price_svm with val_y in mae
mae_svm = mean_absolute_error(val_y, pred_price_svm)
mae_svm

5785.52472

In [34]:
mae_SVM_proc = (mae_svm/y.mean())*100
mae_SVM_proc

42.26198853208474

**Observation: SVM have the worse from previous method MAE equal `5785.52`, representing just `42.26%`.**

In [35]:
comparision_table = pd.DataFrame([
    ['MAE', mae_pred, mae_pred_with_validation, my_mae,mae_random_forest,mae_linear_regression],
    ['MAE of mean %', mae_pred, mae_pred_with_validation_proc, my_mae_proc, mae_random_forest_proc, mae_linear_regression_proc]
], columns = ['','Full fill','Validadion','5000 leaf','Random Forest','Linear Regression'])
comparision_table.round(2)

Full fill  Validadion  5000 leaf  Random Forest  \
0            MAE        0.0     2491.61    2287.94        1879.02   
1  MAE of mean %        0.0       18.20      16.71          13.73   

   Linear Regression  
0            2709.31  
1              19.79